### ISSUE TITLE
Unified Temporal Graph Data Structure

### ISSUE THREAD

**Issue by Jacob-Chmura** [shenyangHuang/openDG#10]
Follow up from: https://github.com/shenyangHuang/openDG/issues/2

### High-level Idea
- Drop the maps and put all the core data in a large pre-allocated **sparse** tensor. 
- A temporal index for time queries
- A topological index for graph queries

Here is a crude depiction of the node store. Analogous tensor would be defined for edges with an extra topological dimension.

![setup](https://github.com/user-attachments/assets/f02fb361-2f92-455a-8f2e-6459728c215f)

**Note**: My suggestion would be to put this entire thing under a private _impl class that we do not expose to clients directly. This is just an implementation, and we should be able to swap it with other (dictionary-based) without any breaking changes to our library.

#### Constructors
```python
def from_csv(csv_file: Pathlib.Path) -> TG:
```

#### Read Ops
```python
def temporal_slice(start_time: Integer, end_time: Integer, materialize: Boolean=False) -> TG | TGView:

def topological_slice(nodes: List[Integer], materialize: Boolean=False) -> TG | TGView:
```

#### Write Ops
```python
def update(e: Event | List[Event]) -> None:

def reindex(dt: TimeGranularity) -> None:
```

#### Accelerator Ops
```python
def to(device: int | str, non_blocking: Boolean=True) -> TG:

def pin_memory() -> TG:

...
```

### Under the hood

Fundamentally, we have a sparse tensor `Storage` primitive:
- `NodeStorage`: $(T \times |V|  \times D_{node})$
- `EdgeStorage`: $(T  \times |V|  \times |V| \times D_{edge})$ -- more likely would go with a sparse topological axis to avoid needed $|V|^2$ adj matrix

And we have an `Index` primitive:
- `TemporalIndex`: Give me time slices from the `Storage`.
- `TopologicalIndex`: Give me topological slices from the `Storage`.

#### Memory
Let's assume, for now, that all the data fits in memory. We need an arena-style memory allocator to pre-alloc a tensor to fit the offline dataset + buffer for online updates. The goal is to minimize memory movement that will affect incoming data.

Once we have our data in working memory, everything becomes a game of efficient indexing and materialization.

#### Indices

For most intents and purposes, we need to support two indexes: a _temporal_ index and a _topological_ index.

**Temporal Index**
_temporal_index[i] = The timestamp of the i'th event in Storage_

The underlying storage is identical across discrete and continuous graphs. Only their temporal indices _may_ differ in interpretation (ints vs. unix epochs). **Converting from discrete to continuous and vice-versa is literally a no-op**.

**Topological Index**
It depends on whether we want node or edge info, but basically, index the inner dimension of the Storage object. 

This data will be scattered across memory, so materializing these views will incur a penalty every time we do it. 

It may be worth looking into [Jagged Tensors](https://pytorch.org/FBGEMM/fbgemm_gpu-overview/jagged-tensor-ops/JaggedTensorOps.html)

#### Views
We will probably want a separate `View` object that can lazily return and apply functions to the `Views` over the Storage before materializing the feature data. 

#### Ops
At a high level, all the ops reduce to indexing the Storage container appropriately:
- _Discretization_: requires merging the storage container using a granularity, which should be translated into an appropriate stride reduction on the underlying storage tensor. The easiest thing to do here is a `torch.stack`
- _Sampling_: Sampling operators can be defined by recursively applying the follow steps in parallel (no need to touch the underlying storage until we have our final index, assuming the sampling methods don't depend on node/edge features)
    - temporal index
    - topological index
    - elementwise function application to temporal index (e.g., time decay weighted)
    - sampling primitive on temporal index
    - Recurse the above for higher-order sampling layers
once we have our final sampled index and materialize the data
- Updates: require in-place updates and buffer management on the temporal allocation end of the storage. Also, we need to be able to grow on the _temporal axis_ as well (e.g., the max number of nodes across all timestamps was 7, but during inference, we see a timestamp with 9 nodes). But since we are dealing with sparse tensors, I think this should be fine.

**Comment by shenyangHuang** [comment]
Guillaume's feedback:
- user should also be able to do naive neighborhood sampling if desired, we can provide option
- conceptually this is the same concept as the other idea of keeping a node based and time based representation, the conceptual operations are the same, this can be how we present the operations to the users in both cases and we can decide on the better backend.
